### PR TITLE
CORE-13863 - extra crypto metrics

### PR DESCRIPTION
**Pull request by conal-smith**
Helm chart changes - https://github.com/corda/engineering-helm/pull/37

### CODE REVIEW

File: components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt [hunk 1]
```diff
@@ -17,7 +17,7 @@ fun getEntityManagerFactory(
     jpaEntitiesRegistry: JpaEntitiesRegistry,
 ): EntityManagerFactory {
     return CordaMetrics.Metric.EntityManagerFactoryCreationTimer.builder()
-        .withTag(CordaMetrics.Tag.VirtualNode, tenantId)
+        .withTag(CordaMetrics.Tag.Tenant, tenantId)
```
Comment (JamesHR3) [inline components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:20]: Is there a distinction between tenant and virtual node?

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt [hunk 2]
```diff
@@ -69,10 +70,22 @@ class CryptoServiceFactoryImpl @Activate constructor(
...
+            .withTag(CordaMetrics.Tag.Method, "FindInstance")
+            .withTag(CordaMetrics.Tag.InstanceType, this::class.java.simpleName)
+            .build()
+            .recordCallable {
+                impl.findInstance(tenantId, category)
+            }!!
 
     override fun getInstance(hsmId: String): CryptoService =
-        impl.getInstance(hsmId)
+        CordaMetrics.Metric.CryptoServiceInstanceCreationTimer.builder()
```
Comment (JamesHR3) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: Any reason for not using the helper you have defined further down for `GetInstance` here?

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt [hunk 3]
```diff
@@ -69,10 +70,22 @@ class CryptoServiceFactoryImpl @Activate constructor(
     )
 
     override fun findInstance(tenantId: String, category: String): CryptoServiceRef =
-        impl.findInstance(tenantId, category)
+        CordaMetrics.Metric.CryptoServiceInstanceCreationTimer.builder()
```
Comment (JamesHR3) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: Should this be the same metric or its own one? Is a new instance being created?

File: libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt [hunk 4]
```diff
@@ -368,6 +368,16 @@ object CordaMetrics {
...
 
+        /**
+         * The time taken to create entity manager factories.
+         */
+        object CryptoServiceInstanceCreationTimer: Metric<Timer>("crypto.service.instance.creation.time", CordaMetrics::timer)
+
+        /**
+         * The time taken to create entity manager factories.
+         */
+        object CryptoMethodTimer: Metric<Timer>("crypto.method.time", CordaMetrics::timer)
```
Comment (JamesHR3) [inline libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt]: One of these kdocs is incorrect.

I have a separate concern that by putting a generic method timer in here it's going to be very hard to determine what exactly is being timed from a user perspective. Should we restrict this to a particular service and tag it by operation for that service?

File: components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt (hunk 1 above)
Comment (kyriathar) [reply components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:20]: So I believe tenant can be virtual node (id) but also "cluster level tenants" like `CRYPTO`, `P2P` and other categories under which keys are added.

File: libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt [hunk 5]
```diff
@@ -571,8 +581,32 @@ object CordaMetrics {
...
 
+        /**
+         * Label to identify the instance of class / implementation retrieved via `getInstance` calls.
+         */
+        InstanceType("instance.type"),
+
+        /**
+         * Label to identify the method inside a class / implementation.
+         */
+        Method("method"),
```
Comment (charlieR3) [inline libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt]: Could you use `OperationName` instead of adding another tag?

File: libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt [hunk 6]
```diff
@@ -571,8 +581,32 @@ object CordaMetrics {
...
+
+        /**
+         * Label to identify the method inside a class / implementation.
+         */
+        PublicKeyType("publickey.type"),
+
+        /**
+         * Label to identify the method inside a class / implementation.
+         */
+        Tenant("tenant"),
```
Comment (charlieR3) [inline libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt]: kdocs are wrong

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt [hunk 7]
```diff
@@ -374,6 +377,11 @@ class SigningServiceImpl(
         if (signingKeyInfo.fullId != requestedFullKeyId) throw IllegalArgumentException(
             "The tenant $tenantId doesn't own public key '${publicKey.publicKeyId()}'."
         )
+        CordaMetrics.Metric.CryptoMethodTimer.builder()
```
Comment (charlieR3) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt]: Could you wrap the whole function in `recordCallable` instead of calculating the duration yourself?

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt [hunk 8]
```diff
@@ -77,6 +82,11 @@ fun CryptoSignatureSpec.toSignatureSpec(serializer: AlgorithmParameterSpecEncodi
         }
         algorithmParams != null -> ParameterizedSignatureSpec(signatureName, algorithmParams)
         else -> SignatureSpecImpl(signatureName)
+    }.also {
+        CordaMetrics.Metric.CryptoMethodTimer.builder()
+            .withTag(CordaMetrics.Tag.Method, CRYPTO_SIGNATURE_SPEC_TO_SIGNATURE_SPEC_METHOD_NAME)
+            .build()
+            .record(Duration.between(startTime, Instant.now()))
```
Comment (charlieR3) [inline libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt:90]: Could this be a recordCallable also?

File: components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt (hunk 1 above)
Comment (sprintengine) [reply components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:20]: Yeah tenant is either a vnodeId or "CRYPTO", "P2P", "REST"

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt (hunk 2 above)
Comment (sprintengine) [reply components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: Hmm it doesn't really feel correct to have a dependency on softhsm-impl for a metric that records getInstance.

Maybe I could find a better home for such a metric, crypto-core perhaps, though not sure I want to add metrics dependency in there.

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt (hunk 7 above)
Comment (sprintengine) [reply components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt]: Slightly changed this around so I've now a "GetOwnedKeyRecordTimer" with label on whether is from "sign" or "deriveSharedSecret" operation.

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt (hunk 8 above)
Comment (sprintengine) [reply libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt:90]: Slightly changed this around so I've now a "CryptoSignatureSpecTimer" with label on operationName (toSignatureSpec or toWire).

I left it in the .also{} at the end because I think it is slightly more readable and I don't need the non-null !! casting.

File: libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt (hunk 6 above)
Comment (sprintengine) [reply libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt]: Thanks yeah I didn't update quite a few, they should all be updated now

File: libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt (hunk 5 above)
Comment (sprintengine) [reply libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt]: Yeah Method is gone, cheers

File: libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt (hunk 4 above)
Comment (sprintengine) [reply libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt]: Yeah I was a bit iffy on this metric, I've split out some of these metrics for example:
- GetOwnedKeyRecordTimer (label on operationName = "deriveSharedSecret" / "sign")
- CryptoCipherSchemeTimer (label on operationName = "decodePublicKeyFromByteArray" / "decodePublicKeyFromString" / "encodePublicKeyToString")
- CryptoSignatureSpecTimer (label on operationName = "toSignatureSpec" / "toWire")

Hopefully this should be a bit better

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt (hunk 3 above)
Comment (sprintengine) [reply components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: I've split out some of the instance creation timers:
- CryptoServiceFindInstanceTimer with label on instance.type ("SoftCryptoServiceImpl")
- CryptoServiceGetInstanceTimer with label on instance.type ("SoftCryptoServiceImpl")
- CryptoSigningRepositoryGetInstanceTimer with label on tenant and instance.type (maybe unnecessary)
- I removed recordGetInstance from softCryptoServiceProvider because this should be wrapped by the `CryptoServiceGetInstanceTimer` already.

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt (hunk 2 above)
Comment (sprintengine) [reply components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: Removed this helper anyway, following below changes

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt [hunk 9]
```diff
@@ -1,5 +1,7 @@
 package net.corda.crypto.service.impl
 
+import java.time.Duration
+import java.time.Instant
```
Comment (kyriathar) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: These imports need be re-ordered I think

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt [hunk 10]
```diff
@@ -68,11 +70,25 @@ class CryptoServiceFactoryImpl @Activate constructor(
...
+                .build()
+                .record(Duration.between(startTime, Instant.now()))
+        }
+    }
 
-    override fun getInstance(hsmId: String): CryptoService =
-        impl.getInstance(hsmId)
+    override fun getInstance(hsmId: String): CryptoService {
+        val startTime = Instant.now()
+        return impl.getInstance(hsmId).also {
```
Comment (kyriathar) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: Not sure about this one. We are adding a metric around something that most of the time will be there in place (`CryptoServiceFactoryImpl.Impl.cryptoService`) or maybe I am missing something (other than the first time which will be initialization through `lazy`). Maybe @dickon can help us here.

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt [hunk 11]
```diff
@@ -68,11 +70,25 @@ class CryptoServiceFactoryImpl @Activate constructor(
         cryptoServiceProvider = cryptoServiceProvider
     )
 
-    override fun findInstance(tenantId: String, category: String): CryptoServiceRef =
-        impl.findInstance(tenantId, category)
+    override fun findInstance(tenantId: String, category: String): CryptoServiceRef {
+        val startTime = Instant.now()
+        return impl.findInstance(tenantId, category).also {
```
Comment (kyriathar) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: Same with `getInstance` I think although this one also involves a DB query, so maybe makes more sense to have a metric. Although I don't think it gets involved in flow-crypto operations.

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt [hunk 12]
```diff
@@ -277,7 +286,15 @@ class SigningServiceImpl(
         otherPublicKey: PublicKey,
         context: Map<String, String>,
     ): ByteArray {
-        val record = getOwnedKeyRecord(tenantId, publicKey)
+        val record =
+            CordaMetrics.Metric.GetOwnedKeyRecordTimer.builder()
+                .withTag(CordaMetrics.Tag.OperationName, DERIVE_SHARED_SECRET_OPERATION_NAME)
```
Comment (kyriathar) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt]: I don't think this operation takes place in flow-crypto interactions either.

File: components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SigningRepositoryFactoryImpl.kt [hunk 13]
```diff
@@ -18,16 +19,23 @@ class SigningRepositoryFactoryImpl(
...
-        layeredPropertyMapFactory = layeredPropertyMapFactory
-    )
+    override fun getInstance(tenantId: String) =
+        CordaMetrics.Metric.CryptoSigningRepositoryGetInstanceTimer.builder()
+            .withTag(CordaMetrics.Tag.InstanceType, SigningRepositoryImpl::class.java.simpleName)
+            .withTag(CordaMetrics.Tag.Tenant, tenantId)
+            .build()
+            .recordCallable {
+                SigningRepositoryImpl(
+                    entityManagerFactory = getEntityManagerFactory(
```
Comment (kyriathar) [inline components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SigningRepositoryFactoryImpl.kt:28]: I think this metric could be a metric around `getEntityManagerFactory`. Rest of it seems static therefore maybe metric noise?

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt [hunk 14]
```diff
@@ -140,31 +145,48 @@ class CipherSchemeMetadataProvider : KeyEncodingService {
...
-    override fun decodePublicKey(encodedKey: ByteArray): PublicKey = try {
-        val subjectPublicKeyInfo = SubjectPublicKeyInfo.getInstance(encodedKey)
-        val scheme = findKeyScheme(subjectPublicKeyInfo.algorithm)
-        val keyFactory = keyFactories[scheme]
-        keyFactory.generatePublic(X509EncodedKeySpec(encodedKey))
-    } catch (e: RuntimeException) {
-        throw e
-    } catch (e: Throwable) {
-        throw CryptoException("Failed to decode public key", e)
+    private fun <T : Any> recordCipherOperation(operationName: String, op: () -> T): T {
```
Comment (kyriathar) [inline libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt]: I don't think there is any cipher operation going on here, is more like serializing/ deserializing keys I think.

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt [hunk 15]
```diff
@@ -77,23 +83,38 @@ fun CryptoSignatureSpec.toSignatureSpec(serializer: AlgorithmParameterSpecEncodi
         }
         algorithmParams != null -> ParameterizedSignatureSpec(signatureName, algorithmParams)
         else -> SignatureSpecImpl(signatureName)
+    }.also {
+        CordaMetrics.Metric.CryptoSignatureSpecTimer.builder()
+            .withTag(CordaMetrics.Tag.OperationName, TO_SIGNATURE_SPEC_OPERATION_NAME)
```
Comment (kyriathar) [inline libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt:88]: Not really sure this metric adds much, I believe it's just Avro signature spec to DTO.

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt [hunk 16]
```diff
@@ -77,23 +83,38 @@ fun CryptoSignatureSpec.toSignatureSpec(serializer: AlgorithmParameterSpecEncodi
...
         is ParameterizedSignatureSpec -> CryptoSignatureSpec(
             signatureName,
             null,
             params.serialize(serializer)
         )
+
         else -> CryptoSignatureSpec(signatureName, null, null)
+    }.also {
+        CordaMetrics.Metric.CryptoSignatureSpecTimer.builder()
+            .withTag(CordaMetrics.Tag.OperationName, TO_WIRE_OPERATION_NAME)
```
Comment (kyriathar) [inline libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt:113]: Same as above, DTO to Avro, seems quite static in terms of time?

**Review by kyriathar — COMMENTED** [review]
Some minor comments, otherwise `sign` and `filterMyKeys` paths are covered.

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt (hunk 15 above)
Comment (sprintengine) [reply libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt:88]: It should include some deserializing too

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt (hunk 16 above)
Comment (sprintengine) [reply libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt:113]: Should include some serializing too

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt (hunk 14 above)
Comment (sprintengine) [reply libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt]: Thanks I renamed this to something slightly better

File: components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SigningRepositoryFactoryImpl.kt (hunk 13 above)
Comment (sprintengine) [reply components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SigningRepositoryFactoryImpl.kt:28]: GetEntityManagerFactory is called in a few places, so at least this gives us detail on time for signing repository.

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt (hunk 12 above)
Comment (sprintengine) [reply components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt]: So `deriveSharedSecret` won't be called during flows? I think for now I might leave this in, and later we can remove a few of these if they prove to be not useful.

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt (hunk 11 above)
Comment (sprintengine) [reply components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: Ok I can remove this findInstance if it isn't relevant. Along with deriveSharedKey call to getOwnedRecords

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt (hunk 10 above)
Comment (sprintengine) [reply components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: Think this one gets called from sign(), and calls the lazy initializer on cryptoService

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt [hunk 17]
```diff
@@ -71,8 +71,9 @@ class CryptoServiceFactoryImpl @Activate constructor(
     override fun findInstance(tenantId: String, category: String): CryptoServiceRef =
         impl.findInstance(tenantId, category)
 
-    override fun getInstance(hsmId: String): CryptoService =
-        impl.getInstance(hsmId)
+    override fun getInstance(hsmId: String): CryptoService {
+        return impl.getInstance(hsmId)
+    }
```
Comment (kyriathar) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt]: Can we please completely revert this change (to keep original git message).

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt [hunk 18]
```diff
@@ -3,6 +3,7 @@ package net.corda.crypto.service.impl
 import com.github.benmanes.caffeine.cache.Cache
 import com.github.benmanes.caffeine.cache.Caffeine
 import java.security.InvalidParameterException
+import java.security.PublicKey
 import java.util.concurrent.TimeUnit
```
Comment (kyriathar) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt:7]: I think this could do with some imports optimizing

File: components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoService.kt [hunk 19]
```diff
@@ -199,33 +200,34 @@ class SoftCryptoService(
...
+            cipher.doFinal(signingData)
+        } else {
+            signatureInstances.withSignature(spec.keyScheme, spec.signatureSpec) { signature ->
+                spec.signatureSpec.getParamsSafely()?.let { params -> signature.setParameter(params) }
+                signature.initSign(privateKey, schemeMetadata.secureRandom)
+                signature.update(signingData)
+                signature.sign()
+            }
+        }
+        CordaMetrics.Metric.Crypto.SoftSignTimer
```
Comment (kyriathar) [inline components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoService.kt]: Perhaps just call `SignTimer`? We don't need "Soft" I think it's confusing (We don't support Hardware HSMs so word "Soft" should just be removed completely at some point).

File: components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoService.kt [hunk 20]
```diff
@@ -199,33 +200,34 @@ class SoftCryptoService(
...
             .builder()
+            .withTag(CordaMetrics.Tag.AlgorithmName, spec.keyScheme.algorithmName)
```
Comment (kyriathar) [inline components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoService.kt]: I think we should rather tag `SigningSpec.signatureSpec` here and not just key's algorithm.

File: components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoServiceProviderImpl.kt [hunk 21]
```diff
@@ -5,6 +5,13 @@ import com.github.benmanes.caffeine.cache.Caffeine
 import com.typesafe.config.ConfigList
 import com.typesafe.config.ConfigObject
 import com.typesafe.config.ConfigValue
+import java.security.InvalidParameterException
+import java.security.KeyPairGenerator
+import java.security.PrivateKey
+import java.security.Provider
+import java.security.PublicKey
+import java.util.InvalidPropertiesFormatException
+import java.util.concurrent.TimeUnit
```
Comment (kyriathar) [inline components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoServiceProviderImpl.kt:14]: imports should be re-ordered here

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt [hunk 22]
```diff
@@ -1,9 +1,16 @@
 package net.corda.crypto.impl
 
+import java.io.StringReader
+import java.io.StringWriter
+import java.security.Provider
+import java.security.PublicKey
+import java.security.SecureRandom
+import java.security.spec.X509EncodedKeySpec
```
Comment (kyriathar) [inline libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt:8]: imports

**Review by kyriathar — COMMENTED** [review]
A few last minor comments

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt (hunk 22 above)
Comment (conalsmith-r3) [reply libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt:8]: Both of these are correct according to my IJ, IJ is a bit inconsistent on the order of these across machines.

File: components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoServiceProviderImpl.kt (hunk 21 above)
Comment (sprintengine) [reply components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoServiceProviderImpl.kt:14]: These appear to be correct according to my IJ settings, what did your optimize imports change?

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt (hunk 22 above)
Comment (sprintengine) [reply libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt:8]: Same here, when you optimize imports what changes here?

**Review by kyriathar — APPROVED** [review]
LGTM

**Review by JamesHR3 — APPROVED** [review]

**Review by rick-r3 — APPROVED** [review]
